src/lib.rs

```diff
@@ -0,0 +1,645 @@
+                        .rem_euclid(table_height as i32)
+                        as usize]
+                },
+                &|x| x.map(|x| -x),
+                &|lhs, rhs| lhs.and_then(|lhs| rhs.map(|rhs| lhs + rhs)),
+                &|lhs, rhs| lhs.and_then(|lhs| rhs.map(|rhs| lhs * rhs)),
+                &|lhs, constant| lhs.map(|lhs| lhs * constant),
+            ),
+            _ => continue,
+        };
+        if let CCSValue::InsideZ(z_index) = cell_mapping.get(&cell_position).copied().unwrap() {
+            if let Some(cell_value) = cell_value {
+                z[z_index] = ARKWORKS::from_le_bytes_mod_order(&cell_value.to_repr());
+            }
+        }
+    }
+
+    z
+}
+
+/// Converts a Halo2 circuit into a sonobe CCS instance.
+///
+/// * `k` log_2(the height of the Plonkish table)
+/// * `c` A Halo2 circuit you wish to convert
+/// * `instance` Assignments to the instance columns. The length of this slice must equal the number of the instance columns in `c`.
+///
+/// Returns a pair of (a ccs instance, the witness vector Z, and lookup constraints)
+/// lookup constraints takes a form of Vec<(L, T)> where for every o ∈ L, z[o] must be in T
+pub fn convert_halo2_circuit<
+    HALO2: ff::PrimeField<Repr = [u8; 32]>,
+    C: Circuit<HALO2>,
+    ARKWORKS: ark_ff::PrimeField,
+>(
+    k: u32,
+    circuit: &C,
+    instance: &[&[HALO2]],
+) -> Result<
+    (
+        CCS<ARKWORKS>,
+        Vec<ARKWORKS>,
+        Vec<(HashSet<usize>, HashSet<ARKWORKS>)>,
+    ),
+    plonk::Error,
+> {
+    let mut meta = ConstraintSystem::<HALO2>::default();
+    let config = C::configure(&mut meta);
+
+    // both AssignmentDumper and generate_ccs_instance expects Vec of length 2^k
+    // so I need to fill the vec
+    let instance_option: Vec<Vec<Option<HALO2>>> = instance
+        .iter()
+        .map(|cells| {
+            let mut column = vec![None; 1 << k];
+            column[0..cells.len()]
+                .copy_from_slice(&cells.iter().map(|cell| Some(*cell)).collect::<Vec<_>>());
+            column
+        })
+        .collect();
+    let instance_value: Vec<Vec<Value<HALO2>>> = instance
+        .iter()
+        .map(|cells| {
+            let mut column = vec![Value::unknown(); 1 << k];
+            column[0..cells.len()].copy_from_slice(
+                &cells
+                    .iter()
+                    .map(|cell| Value::known(*cell))
+                    .collect::<Vec<_>>(),
+            );
+            column
+        })
+        .collect();
+
+    let mut cell_dumper: AssignmentDumper<HALO2> = AssignmentDumper::new(k, &meta);
+    cell_dumper.instance = instance_value;
+    C::FloorPlanner::synthesize(&mut cell_dumper, circuit, config, meta.constants.clone())?;
+
+    // FIXME: Isn't there a better way to make &[&[A]] from Vec<Vec<A>>?
+    let instance_option: Vec<&[Option<HALO2>]> = instance_option
+        .iter()
+        .map(|x| x.as_slice())
+        .collect::<Vec<_>>();
```

Comment:
You can remove type inference here as well! (and same for below)
```suggestion
        .collect();
```